File: Python/Searching and Sorting/18_minimum_swaps_to_sort.py
"""
Idea:
using hashmap to store indices of elements in input vector & copying the vector & sort it for reference
Steps:
iterating each element, if an element is not in position, swap it using hashmap & sorted vector.
"""


class Solution:

    def minSwaps(self, nums):
        n = len(nums)
        sortedArray = nums[:]
        sortedArray.sort()  # sorting the copy of input vector
        hashMap = {}
        for i in range(n):
            if nums[i] not in hashMap:  # creating hashmap
                hashMap[nums[i]] = i
            else:
                hashMap[nums[i]] += 1
        ans = 0
        for i in range(n):
            if nums[i] != sortedArray[i]:  # if element is not in optimum position
                ans += 1
                s, t = nums[i], sortedArray[i]
                nums[i], nums[hashMap[sortedArray[i]]] = nums[hashMap[sortedArray[i]]], nums[i]  # swapping the element using hashmap
                hashMap[s] = hashMap[t]  # updating the hashmap

        return ans


if __name__ == '__main__':
    T = int(input())
    for i in range(T):
        n = int(input())
        nums = list(map(int, input().split()))
        obj = Solution()
        ans = obj.minSwaps(nums)
        print(ans)

# } Driver Code Ends

File: C++/Searching and Sorting/18_minimum_swaps_to_sort.cpp

/*
Idea:
using hashmap to store indices of elements in input vector & copying the vector & sort it for reference
Steps:
iterating each element, if an element is not in position, swap it using hashmap & sorted vector.
*/

#include<bits/stdc++.h>
using namespace std;



class Solution
{
public:
    int minSwaps(vector<int>& nums)
    {
        int n = nums.size(), ans = 0;
        vector<int> sortedArray = nums;
        sort(sortedArray.begin(), sortedArray.end()); // sorting the copy of input vector
        unordered_map<int, int>hashMap;

        for (int i = 0; i < n; i++)
            hashMap[nums[i]] = i; // creating hashmap

        for (int i = 0;i < n;++i)
            if (nums[i] != sortedArray[i]) // if element is not in optimum position
            {
                ans++;
                int s = nums[i], t = sortedArray[i];
                swap(nums[i], nums[hashMap[sortedArray[i]]]); // swapping the element using hashmap
                swap(hashMap[s], hashMap[t]); // updating the hashmap
            }
        return ans;
    }
};

int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int n;
        cin >> n;
        vector<int>nums(n);
        for (int i = 0; i < n; i++)
            cin >> nums[i];
        Solution obj;
        int ans = obj.minSwaps(nums);
        cout << ans << "\n";
    }
    return 0;
}